File: Cpp/1143.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        vector<vector<int>> arr(text1.length(), vector<int>(text2.length(), 0));
        for(int i = 0; i < text1.length(); i++){
            if(text1[i] == text2[0] || (i != 0 && arr[i-1][0] == 1)){arr[i][0] = 1;}
        }
        for(int j = 0; j < text2.length(); j++){
            if(text1[0] == text2[j] || (j != 0 && arr[0][j-1] == 1)){arr[0][j] = 1;}
        }
        for(int i = 1; i < text1.length(); i++){
            for(int j = 1; j < text2.length(); j++){
                if(text1[i] == text2[j]){
                    arr[i][j] = arr[i-1][j-1]+1;
                }else{
                    arr[i][j] = max(arr[i-1][j], arr[i][j-1]);
                }
            }
        }
        return arr.back().back();
    }
};

File: Python/1143.py
class Solution:
    def f(self, i, j, dp, text1, text2):
        
        if i >= len(text1) or j >= len(text2):
            return 0
        if (i,j) in dp:
            return dp[(i,j)]
        if text1[i] == text2[j]:
            dp[(i,j)] = 1 + self.f(i+1,j+1, dp, text1, text2)
            return dp[(i,j)]
        else:
            dp[(i,j)] = max(self.f(i,j+1, dp, text1, text2), self.f(i+1,j, dp, text1, text2))
            return dp[(i,j)]


    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = {}
        return self.f(0,0,dp,text1, text2)
